test/beast/core/test_buffer.hpp

```diff
@@ -593,6 +598,224 @@ test_dynamic_buffer(
+        b.shrink(shrink_size);
+        expected = expected.substr(0, shrink_size > expected.size() ? 0 : expected.size() - shrink_size);
+        BEAST_EXPECT(store.size() == expected.size());
+        BEAST_EXPECT(b.size() == expected.size());
+        BEAST_EXPECT(buffers_to_string(store.data()) == expected);
+        BEAST_EXPECT(buffers_to_string(b.data(0, expected.size())) == expected);
+    };
+
+    transition_test(0, 128, 0, 128, 0);
+
+    for (auto residual_front : testing::intervals(0, 1024, 256))
+    for (auto initial_size : testing::intervals(0, initial_data.size(), 128))
+    for (auto extra_prep : testing::intervals(0, 712, 256))
+    for (auto grow_size : testing::intervals(0, added_data.size(), 128))
+    for (auto shrink_size : testing::intervals(0, grow_size + initial_size, 128))
+    transition_test(residual_front, initial_size, extra_prep, grow_size, shrink_size);
+}
+
+template<class DynamicBuffer_v0>
+void test_dynamic_buffer_v0_v2_consistency()
+{
+    test_dynamic_buffer_v0_v2_consistency([]{ return DynamicBuffer_v0(); });
+}
+
+template<class DynamicBuffer_v0>
+void
+test_dynamic_buffer_v0_v2_operation(DynamicBuffer_v0 storage)
+{
+    BOOST_ASSERT(storage.max_size() == 16);
+    BOOST_ASSERT(storage.size() == 0);
+
+    using storage_type = decltype(storage);
+
+    BEAST_EXPECT(detail::is_dynamic_buffer_v0<storage_type>::value);
+
+    auto dyn_buf = dynamic_buffer(storage);
+
+
+    BEAST_EXPECT(dyn_buf.size() < dyn_buf.max_size());
+    BEAST_EXPECT(dyn_buf.size() == 0);
+    BEAST_EXPECT(net::buffer_size(dyn_buf.data(0, dyn_buf.size())) == 0);
+
+    auto do_insert = [&dyn_buf](net::const_buffer source)
+    {
+        auto start = dyn_buf.size();
+        auto len = source.size();
+        dyn_buf.grow(len);
+        auto insert_region = dyn_buf.data(start, len);
+        BEAST_EXPECT(net::buffer_size(insert_region) == len);
+        auto copied = net::buffer_copy(insert_region, source);
+        BEAST_EXPECT(copied == len);
+    };
+
+    do_insert(net::buffer(std::string("0123456789")));
+    dyn_buf.shrink(1);
+    auto output_region = dyn_buf.data(0, dyn_buf.size());
+    BEAST_EXPECT(net::buffer_size(output_region) == 9);
+    BEAST_EXPECT(buffers_to_string(output_region) == "012345678");
+
+    do_insert(net::buffer(std::string("9abcdef")));
+    dyn_buf.shrink(0);
+    output_region = dyn_buf.data(0, dyn_buf.size());
+    BEAST_EXPECT(net::buffer_size(output_region) == 16);
+    BEAST_EXPECT(buffers_to_string(output_region) == "0123456789abcdef");
+
+    BEAST_THROWS(dyn_buf.grow(1), std::length_error);
+
+    dyn_buf.consume(10);
+    output_region = dyn_buf.data(0, dyn_buf.size());
+    BEAST_EXPECT(net::buffer_size(output_region) == 6);
+    BEAST_EXPECT(buffers_to_string(output_region) == "abcdef");
+
+    dyn_buf.consume(10);
+    output_region = dyn_buf.data(0, dyn_buf.size());
+    BEAST_EXPECT(net::buffer_size(output_region) == 0);
+}
+
+template<class DynBufferV0Generator>
```

Comment:
Too many new names! "Dyn" now? What's wrong with `template< class G >`

Reply:
I prefer descriptive names. One day this will be a concept. But sure, not problem to change it.